### PR TITLE
isabelle: Prebuild HOL session

### PR DESCRIPTION
**Pull request by jvanbruegge**
Upstream isabelle ships with this session, but the changes done for
nixpkgs invalidate it, requiring every user to build it again.

###### Motivation for this change

#154249 
This also ensures that future updates can at least build the HOL session

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jvanbruegge** [comment]
Ibsee there was an error in the build, do not merge this yet. I will fix it later

**Comment by jvanbruegge** [comment]
Fixed the issues now

**Comment by jvanbruegge** [comment]
And resolved the merge conflict

**Comment by jvanbruegge** [comment]
But now, the build failure is back

**Comment by jvanbruegge** [comment]
Ok, it is not really needed to fix this exception in the build. The heaps are still generated. So I'd say, merge this and I can fix this issue in the future by figuring out how to build `isabelle_setup` from source

**Comment by jvanbruegge** [comment]
cc @jwiegley @gebner

**Comment by gebner** [comment]
During build, I get this error:
```
Exception in thread "Isabelle.thread-11" Exception in thread "Isabelle.thread-15" java.io.IOException: Cannot run program "/usr/bin/env": error=2, No such file or directory
        at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1143)
        at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1073)
        at isabelle.setup.Environment.exec_process(Environment.java:234)
        at isabelle.setup.Environment.kill_process(Environment.java:451)
        at isabelle.Bash$Process.signal(bash.scala:130)
        at isabelle.Bash$Process.$anonfun$terminate$1(bash.scala:144)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
        at isabelle.Isabelle_Thread.interrupt_handler(isabelle_thread.scala:188)
        at isabelle.Isabelle_Thread$.interrupt_handler(isabelle_thread.scala:116)
        at isabelle.Isabelle_Thread$.try_uninterruptible(isabelle_thread.scala:128)
        at isabelle.Bash$Process.terminate(bash.scala:143)
        at isabelle.Bash$Process.$anonfun$shutdown_hook$1(bash.scala:157)
        at isabelle.Isabelle_Thread.run(isabelle_thread.scala:141)
Caused by: java.io.IOException: error=2, No such file or directory
        at java.base/java.lang.ProcessImpl.forkAndExec(Native Method)
        at java.base/java.lang.ProcessImpl.<init>(ProcessImpl.java:314)
        at java.base/java.lang.ProcessImpl.start(ProcessImpl.java:244)
        at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1110)
        ... 12 more
java.io.IOException: Cannot run program "/usr/bin/env": error=2, No such file or directory
        at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1143)
        at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1073)
        at isabelle.setup.Environment.exec_process(Environment.java:234)
        at isabelle.setup.Environment.kill_process(Environment.java:451)
        at isabelle.Bash$Process.signal(bash.scala:130)
        at isabelle.Bash$Process.$anonfun$terminate$1(bash.scala:144)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
        at isabelle.Isabelle_Thread.interrupt_handler(isabelle_thread.scala:188)
        at isabelle.Isabelle_Thread$.interrupt_handler(isabelle_thread.scala:116)
        at isabelle.Isabelle_Thread$.try_uninterruptible(isabelle_thread.scala:128)
        at isabelle.Bash$Process.terminate(bash.scala:143)
        at isabelle.Bash$Process.$anonfun$shutdown_hook$1(bash.scala:157)
        at isabelle.Isabelle_Thread.run(isabelle_thread.scala:141)
Caused by: java.io.IOException: error=2, No such file or directory
        at java.base/java.lang.ProcessImpl.forkAndExec(Native Method)
        at java.base/java.lang.ProcessImpl.<init>(ProcessImpl.java:314)
        at java.base/java.lang.ProcessImpl.start(ProcessImpl.java:244)
        at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1110)
        ... 12 more
```

The build continues though, and the prebuilt HOL session seems to work fine.  IIRC a long time ago we prebuilt the HOL session, but I've disabled it in 2016 because it didn't seem to work (the HOL session was still getting built on first startup).  It's great to see this working again!

**Comment by wizeman** [comment]
FYI, there's been previous work on this, with some minor differences, e.g.: https://github.com/NixOS/nixpkgs/pull/115339/commits/9933d71ea52d7b92a623ee2c2317a5fbf3b00108, part of PR #115339.

But apart from the apparently unneeded `bash` dependency it LGTM.

**Comment by jvanbruegge** [comment]
The approach of #115339 does not work any more with Isabelle2021-1. To fix the error it would require recompiling the `isabelle_setup` component and I have not found out how. The HOL build needs the bash input.

**Comment by wizeman** [comment]
> The approach of #115339 does not work any more with Isabelle2021-1. To fix the error it would require recompiling the `isabelle_setup` component and I have not found out how.

I think you might be right, I tried forward porting that commit but also ran into a build failure.

> The HOL build needs the bash input.

That might be true, but you're not actually adding bash as an input, though. You're only adding it as a derivation parameter, but you're not using it anywhere in the derivation, so the parameter is not necessary.

I think `stdenv` already includes bash anyway, so there should be no need to add bash as a `buildInput`.

**Comment by jvanbruegge** [comment]
> I think you might be right, I tried forward porting that commit but also ran into a build failure.

The handling of external processes changed in this release, that's why it is not working any more.

I also removed the bash input and verified that the heaps that are built are indeed working as expected. The error can be safely ignored

**Comment by jvanbruegge** [comment]
In case someone finds this PR and wonders why Isabelle still rebuilds the `HOL` heap image: **Delete your `$HOME/.isabelle/heaps`** directory. If you used an earlier version of the package, isabelle built the heap image for HOL and Pure there. As the heaps in `$HOME/.isabelle` are preferred to the one in the isabelle install directory, and the heaps there are out of date, isabelle rebuilds them.

TL;DR: Delete `$HOME/.isabelle/heaps`, will avoid rebuilding HOL yourself
